crates/sui-core/src/unit_tests/authority_tests.rs

```diff
@@ -5298,6 +5298,124 @@ async fn test_for_inc_201_dry_run() {
     assert_eq!(json!({"foo":"bar"}), events.data[0].parsed_json);
 }
 
+#[tokio::test]
+async fn test_publish_transitive_dependencies_ok() {
+    use sui_framework_build::compiled_package::BuildConfig;
+
+    let (sender, key): (_, AccountKeyPair) = get_key_pair();
+    let gas_id = ObjectID::random();
+    let state = init_state_with_ids(vec![(sender, gas_id)]).await;
+
+    // Get gas object
+    let gas_object = state.get_object(&gas_id).await.unwrap().unwrap();
+    let gas_ref = gas_object.compute_object_reference();
+
+    // Publish `package C`
+    let mut package_c_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+    package_c_path.extend([
+        "src",
+        "unit_tests",
+        "data",
+        "transitive_dependencies",
+        "package_C",
+    ]);
+    let modules = BuildConfig::new_for_testing()
+        .build(package_c_path)
+        .unwrap()
+        .get_package_bytes(/* with_unpublished_deps */ false);
+
+    let mut builder = ProgrammableTransactionBuilder::new();
+    builder.publish_immutable(modules, vec![]);
+    let kind = TransactionKind::programmable(builder.finish());
+    let txn_data = TransactionData::new_with_gas_coins(kind, sender, vec![gas_ref], 10000, 1);
+    let signed = to_sender_signed_transaction(txn_data, &key);
+    let txn_effects = send_and_confirm_transaction(&state, signed)
+        .await
+        .unwrap()
+        .1
+        .into_data();
+    let (package_c_id, _, _) = &txn_effects.created()[0].0;
+
+    // Publish `package B`
+    let mut package_b_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+    package_b_path.extend([
+        "src",
+        "unit_tests",
+        "data",
+        "transitive_dependencies",
+        "package_B_depends_on_C",
+    ]);
+    let modules = BuildConfig::new_for_testing()
+        .build(package_b_path)
+        .unwrap()
+        .get_package_bytes(/* with_unpublished_deps */ false);
+
+    let mut builder = ProgrammableTransactionBuilder::new();
+
+    builder.publish_immutable(modules, vec![*package_c_id]); // Note: B depends on C
+
+    let kind = TransactionKind::programmable(builder.finish());
+    let txn_data = TransactionData::new_with_gas_coins(kind, sender, vec![gas_ref], 10000, 1);
+    let signed = to_sender_signed_transaction(txn_data, &key);
+    let txn_effects = send_and_confirm_transaction(&state, signed)
+        .await
+        .unwrap()
```

Comment:
Running into:

```
panicked at 'called `Result::unwrap()` on an `Err` value: UserInputError { error: 
ObjectVersionUnavailableForConsumption { provided_obj_ref:
(0xe8d6f034edc5f165ba1ca41d96ad770035332dbbfbceb1f85bb0eecdd36a611f, SequenceNumber(1),
 o#71T1BA9JmVkpoLiuZUugdWDq6WycVydVHwidHrpG9aCs), current_version: SequenceNumber(2)
} }', crates/sui-core/src/unit_tests/authority_tests.rs:5362:10
```

This is the second block that tries to repeat publishing a second package. I'm clearly doing something silly, but not sure what. Is it to do with `gas` and/or reusing `gas_ref`?

Reply:
Yep, exactly right, it's `gas_ref` -- `ObjectRef`'s include the object version (and digest), and each version can only be used for one transaction, so you will need to find the object ref of the new gas object from the effects of your previous transaction -- there should be a special field in effects that is specifically for the gas object.